dc/s2n-quic-dc/src/uds/receiver.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use nix::{
+    fcntl::{fcntl, FcntlArg, OFlag},
+    sys::socket::{
+        bind, recvmsg, socket, AddressFamily, ControlMessageOwned, MsgFlags, SockFlag, SockType,
+        UnixAddr,
+    },
+    unistd::{close, unlink},
+};
+use std::{
+    os::{
+        fd::AsFd,
+        unix::io::{AsRawFd, RawFd},
+    },
+    path::{Path, PathBuf},
+};
+use tokio::io::{unix::AsyncFd, Interest};
+
+const BUFFER_SIZE: usize = u16::MAX as usize;
+
+pub struct Receiver {
+    async_fd: AsyncFd<RawFd>,
+    socket_path: PathBuf,
+}
+
+impl Receiver {
+    pub fn new(socket_path: &Path) -> Result<Self, std::io::Error> {
+        // Remove existing socket file if it exists
+        unlink(socket_path)?;
+
+        let socket_owned = socket(
+            AddressFamily::Unix,
+            SockType::Datagram,
+            SockFlag::empty(),
+            None,
+        )?;
+        let socket_fd = socket_owned.as_raw_fd();
+
+        // Set socket to non-blocking mode
+        let flags = fcntl(socket_owned.as_fd(), FcntlArg::F_GETFL)?;
+        let new_flags = OFlag::from_bits_truncate(flags) | OFlag::O_NONBLOCK;
+        fcntl(socket_owned.as_fd(), FcntlArg::F_SETFL(new_flags))?;
+
+        let unix_addr = UnixAddr::new(socket_path)?;
+        bind(socket_fd, &unix_addr)?;
+
+        // Prevent the socket from being closed when socket_owned goes out of scope
+        std::mem::forget(socket_owned);
```

Comment:
This seems like it's going to leak the socket? I think you want `AsyncFd<std::os::fd::OwnedFd>` perhaps?